lib/debug.py

```diff
@@ -0,0 +1,197 @@
+        })
+        self.svg.append(layer)
+        self.current_layer = layer
+
+    @check_enabled
+    def open_group(self, name="Group"):
+        group = etree.Element("g", {
+            INKSCAPE_LABEL: name
+        })
+
+        self.log_svg_element(group)
+        self.group_stack.append(group)
+
+    @check_enabled
+    def close_group(self):
+        if self.group_stack:
+            self.group_stack.pop()
+
+    @check_enabled
+    def log(self, message, *args):
+        if self.last_log_time:
+            message = "(+%s) %s" % (datetime.now() - self.last_log_time, message)
+
+        self.raw_log(message, *args)
+
+    def raw_log(self, message, *args):
+        now = datetime.now()
+        timestamp = now.isoformat()
+        self.last_log_time = now
+
+        print >> self.log_file, timestamp, message % args
+        self.log_file.flush()
+
+    def time(self, func):
+        def decorated(*args, **kwargs):
+            if self.enabled:
+                self.raw_log("entering %s()", func.func_name)
+                start = time.time()
+
+            result = func(*args, **kwargs)
+
+            if self.enabled:
+                end = time.time()
+                self.raw_log("leaving %s(), duration = %s", func.func_name, round(end - start, 6))
+
+            return result
+
+        return decorated
+
+    @check_enabled
+    def log_svg_element(self, element):
+        if self.current_layer is None:
+            self.add_layer()
+
+        if self.group_stack:
+            self.group_stack[-1].append(element)
+        else:
+            self.current_layer.append(element)
+
+    @check_enabled
+    def log_line_string(self, line_string, name=None, color=None):
+        """Add a Shapely LineString to the SVG log."""
+        self.log_line_strings([line_string], name, color)
+
+    @check_enabled
+    def log_line_strings(self, line_strings, name=None, color=None):
+        path = line_strings_to_path(line_strings)
+        path.set('style', formatStyle({"stroke": color or "#000000", "stroke-width": "0.3"}))
+
+        if name is not None:
+            path.set(INKSCAPE_LABEL, name)
+
+        self.log_svg_element(path)
+
+    @check_enabled
+    def log_line(self, start, end, name="line", color=None):
+        self.log_svg_element(etree.Element("path", {
+            "d": "M%s,%s %s,%s" % (start + end),
+            "style": formatStyle({"stroke": color or "#000000", "stroke-width": "0.3"}),
+            INKSCAPE_LABEL: name
+        }))
+
+    @check_enabled
+    def log_graph(self, graph, name="Graph", color=None):
+        d = ""
+
+        for edge in graph.edges:
+            d += "M%s,%s %s,%s" % (edge[0] + edge[1])
+
+        self.log_svg_element(etree.Element("path", {
+            "d": d,
+            "style": formatStyle({"stroke": color or "#000000", "stroke-width": "0.3"}),
+            INKSCAPE_LABEL: name
+        }))
+
+    @contextmanager
+    def time_this(self, label="code block"):
```

Comment:
does this context ever get applied when self.enabled is False?

Reply:
Yup!  My idea with this class is that I can add debug statements to the code and leave them in.  When debugging is disabled, the debug statements should have minimal impact.  For this context manager, it should just be a noop if debug is disabled.

Previously I was adding and removing debug statements whenever I needed them.  Especially for things like timing code, it was getting tedious.